src/Symfony/Component/Validator/Constraints/NoBannedWordsValidator.php

```diff
@@ -0,0 +1,92 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Validator\Constraints;
+
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+use Symfony\Component\Validator\Exception\UnexpectedTypeException;
+use Symfony\Component\Validator\Exception\UnexpectedValueException;
+
+final class NoBannedWordsValidator extends ConstraintValidator
+{
+    public function validate(#[\SensitiveParameter] mixed $value, Constraint $constraint): void
+    {
+        if (!$constraint instanceof NoBannedWords) {
+            throw new UnexpectedTypeException($constraint, NoBannedWords::class);
+        }
+
+        if (null === $value || !$constraint->dictionary) {
+            return;
+        }
+
+        if (!\is_string($value)) {
+            throw new UnexpectedValueException($value, 'string');
+        }
+
+        $matches = [];
+        $regex = sprintf(
+            '{%s}i',
+            implode('|', array_map($this->toL33tRegex(...), $constraint->dictionary))
+        );
+
+        $found = preg_match_all($regex, $value, $matches);
+        if (false === $found || 0 === $found) {
+            return;
+        }
+        $matches = current($matches);
+
+        if ($matches) {
+            $this->context->buildViolation($constraint->message, [
+                '{{ matches }}' => implode(', ', $matches),
+                '{{ dictionary }}' => implode(', ', $constraint->dictionary),
+            ])
+                ->setCode(NoBannedWords::BANNED_WORDS_ERROR)
+                ->addViolation();
+        }
+    }
+
+    private function toL33tRegex(string $data): string
+    {
+        $partsToRegex = fn (array $parts): string => sprintf('(%s)', implode('|', array_map(preg_quote(...), array_unique($parts))));
+
+        return implode('', array_map(function (string $char) use ($partsToRegex): string {
+            return strtr($char, [
+                'a' => $partsToRegex(['a', '4', '/\\', '@', '^', 'aye', '∂', '/-\\', '|-\\', 'q']),
```

Comment:
let's inline this map, there's no need to compute it at runtime at every iteration
this while method would be replaced by a single call to strtr with a pre-computed map